app-backend/tile/src/main/scala/image/Mosaic.scala

```diff
@@ -36,123 +38,88 @@ object Mosaic {
       } catch {
         case e: AttributeNotFoundError if tryZoom > 0 => readMetadata(store, tryZoom - 1)
       }
 
-    tileLayerMetadataCache.get(s"mosaic-tlm-$id-$zoom", { cKey =>
-      for {
-        prefix <- LayerCache.prefixFromLayerId(id)
-        store <- LayerCache.attributeStore(prefix)
-      } yield {
+    memcached.caching(s"mosaic-tlm-$id-$zoom") { _ =>
+      OptionT(LayerCache.attributeStoreForLayer(id)).mapFilter { store =>
         readMetadata(store, zoom)
-      }
-    })
+      }.value
+    }
   }
 
-
-  val mosaicDefinitionCache = HeapBackedMemcachedClient[Option[MosaicDefinition]](memcachedClient)
-  def mosaicDefinition(projectId: UUID, tagttl: Option[TagWithTTL])(implicit db: Database) = {
+  def mosaicDefinition(projectId: UUID, tagttl: Option[TagWithTTL])(implicit db: Database): Future[Option[MosaicDefinition]] = {
+    // TODO: TTL in tagttl does nothing, default memcachd TTL is used anyway
     val cacheKey = tagttl match {
       case Some(t) => s"mosaic-definition-$projectId-${t.tag}"
       case None => s"mosaic-definition-$projectId"
     }
-    mosaicDefinitionCache.caching(cacheKey) { implicit ec =>
+
+    memcached.caching(cacheKey) { _ =>
       ScenesToProjects.getMosaicDefinition(projectId)
     }
   }
 
   /** Fetch the tile for given resolution. If it is not present, use a tile from a lower zoom level */
-  def fetch(id: UUID, zoom: Int, col: Int, row: Int)(implicit database: Database): Future[Option[MultibandTile]] =
-    tileLayerMetadata(id, zoom).flatMap {
-      case Some((sourceZoom, tlm)) =>
-        val zdiff = zoom - sourceZoom
-        val pdiff = 1 << zdiff
-        val sourceKey = SpatialKey(col / pdiff, row / pdiff)
-        if (tlm.bounds includes sourceKey)
-          for ( maybeTile <- LayerCache.maybeTile(id, sourceZoom, sourceKey) ) yield {
-            for (tile <- maybeTile) yield {
-              val innerCol  = col % pdiff
-              val innerRow  = row % pdiff
-              val cols = tile.cols / pdiff
-              val rows = tile.rows / pdiff
-              tile.crop(GridBounds(
-                colMin = innerCol * cols,
-                rowMin = innerRow * rows,
-                colMax = (innerCol + 1) * cols - 1,
-                rowMax = (innerRow + 1) * rows - 1
-              )).resample(256, 256)
-            }
-          }
-        else
-          Future.successful(None)
-      case None =>
-        Future.successful(None)
+  def fetch(id: UUID, zoom: Int, col: Int, row: Int)(implicit database: Database): Future[Option[MultibandTile]] = {
+    for {
+      (sourceZoom, tlm) <- OptionT(tileLayerMetadata(id, zoom))
+      zoomDiff = zoom - sourceZoom
+      resolutionDiff = 1 << zoomDiff
+      sourceKey = SpatialKey(col / resolutionDiff, row / resolutionDiff)
+      tile <- OptionT(LayerCache.layerTile(id, sourceZoom, sourceKey)) if tlm.bounds.includes(sourceKey)
```

Comment:
It's been my experience with Monad Transformers that adopting a "unwrap as late as possible" approach is generally beneficial. Just like when we have a lot of functions that return `Future` we say to ourselves "I'm working in the `Future` context", we can think with `OptionT` like "I'm in the `OptionT` context". If we make that assumption, we gain some power, namely that we can `flatMap` through these operations with no other overhead, and expect the chained operations to fail if any step held an internal `None`.

```scala
def foo: OptionT[Future, A]
def bar(a: A): OptionT[Future, B]
def baz(b: B): OptionT[Future, C]

def work: OptionT[Future, C] = foo.flatMap(bar).flatMap(baz)  // foo >>= bar >>= baz
```
as opposed to the pattern here, where you've had to rewrap `LayerCache.layerTile` even though it started in an `OptionT`. 

The rule of thumb then could be: "work in `OptionT[Future, A]` as long as possible until some caller actually needs only a `Future[A]`. Then unwrap the `OptionT` once, pattern match on the result, and you're off to the races."

I'll generally assert that provided one knows what a Monad Transformer is, `OptionT[Future, A]` provides more information *and* power than `Future[Option[A]]`.